### PR TITLE
feat: construct base class changes (breaking)

### LINKED ISSUE #189: Rename construct's "name" parameter to "id" or "cid" or "conid"
Feedback from @skinny85, which I agree. The "name" terminology is easily confused with the physical name of the resource. We should use `id`.

See #1423 - we should consider perhaps to give IDs a more distinctive name such as "cid" or "conid" so that people will understand that these have a few limitations (for example, you can't use tokens).

### LINKED ISSUE #1431: construct: consider renaming "parent" to "scope"
I am wondering if maybe some of the confusion and non-intuitiveness around the construct tree can be relieved by modifying the semantics ever so slightly so they will reflect more accurately the reason behind the first constructor argument. Perhaps instead of calling it "parent" (which represents the in-memory data structure) we should call it "scope" (which more accurately represents the motivation behind the model).

I have a feeling that by calling the first argument "scope", it would be more natural for users to pass in "this", indicating that this object is created within *this scope*.

Another benefit of this is that it will reduce the temptation to overlay other tree models (e.g. route53, step functions, pipelines) onto the construct hierarchy, since the mental model would not align so cleanly.

The underlying mechanisms will be exactly the same.

### LINKED ISSUE #1441: constructs: reduce API clutter from base class
The surface area of the `Construct` base class is polluting the surface area of derived classes without adding much value to the derived classes' users (in most cases, users won't need to interact with the construct surface).

One of the ideas that came up is grouping most of this surface area under some property (e.g. `construct.node` so that the surface area will not be cluttered.

### PR DESCRIPTION
**Pull request by eladb**
- Rename "parent" to "scope" in all construct initializers and imports (fixes #1431)
- Group all members (besides `validate`) under a `node` property so they don't pollute the surface area of derived classes (fixes #1441)
- Define `IConstruct` and have all `IXxx` extend it

BREAKING CHANGE: Multiple breaking changes:

  - All the APIs of `cdk.Construct` are now available under `construct.node` (e.g. instead of `construct.path` use `construct.node.path`. See #1441 for details.
  - Construct `parent` was renamed to `scope`. See #1431 for more details.
  - First argument of `codepipeline.CrossRegionScaffoldStack` is now required.
  - `IPipeline.uniqueId` removed and can now be accessed via `pipeline.node.uniqueId`.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added
  - __CLI change?:__ manually run [integration tests](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/integ-tests/test.sh) and paste output as a PR comment
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/cdk/lib/core/construct.ts [hunk 1]
```diff
@@ -2,22 +2,18 @@ import cxapi = require('@aws-cdk/cx-api');
 import { makeUniqueId } from '../util/uniqueid';
 export const PATH_SEP = '/';
 
-/**
- * Represents the building block of the construct graph.
- * When a construct is created, it is always added as a child
- */
-export class Construct {
+export class Node {
```
Comment (rix0rrr) [inline packages/@aws-cdk/cdk/lib/core/construct.ts]: Can't help but feel that `Node` is overly generic. `ConstructNode` maybe, or even `ConstructTree`?

Other than that, muy bien!

**Comment by eladb** [comment]
@rix0rrr @RomainMuller I am loving "scope", but not 100% comfortable with renaming "id" to "scid". How do you feel about it?

**Comment by eladb** [comment]
The reason it feels wrong is because we are starting to see a pattern where this ID is used as a double for some other lower-level name/ID. For example, the name of a CodePipeline stage, the name of the Stack, ID of the Step Functions action, etc.

On the other hand, due to the fact that these IDs cannot include deploy-time tokens, I am worried that people are going to abuse them and constantly fall into this error, so perhaps we _should_ give them a wacky name so they will less likely be abused.

File: packages/@aws-cdk/cdk/lib/core/construct.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/core/construct.ts]: I would support both `treeNode` and `node`.

**Comment by rix0rrr** [comment]
I'm fine with leaving `id`, as well as `scid`.

**Pull request merged**